cmd/api/src/database/migration/migrations/v6.1.0.sql

```diff
@@ -14,29 +14,84 @@
+    id              SERIAL PRIMARY KEY,
+    client_id       TEXT                                                    NOT NULL,
+    issuer          TEXT                                                    NOT NULL,
+    sso_provider_id INTEGER REFERENCES sso_providers (id) ON DELETE CASCADE NULL,
 
-  UNIQUE (name)
+    updated_at      TIMESTAMP WITH TIME ZONE DEFAULT now(),
+    created_at      TIMESTAMP WITH TIME ZONE DEFAULT now()
 );
 
+-- Create the reference from saml_providers to sso_providers
+ALTER TABLE ONLY saml_providers
+    ADD COLUMN IF NOT EXISTS sso_provider_id INTEGER NULL;
+ALTER TABLE ONLY saml_providers
+    DROP CONSTRAINT IF EXISTS fk_saml_provider_sso_provider;
+ALTER TABLE ONLY saml_providers
+    ADD CONSTRAINT fk_saml_provider_sso_provider FOREIGN KEY (sso_provider_id) REFERENCES sso_providers (id) ON DELETE CASCADE;
 
--- Add Scheduled Analysis Configs
-INSERT INTO parameters (key, name, description, value, created_at, updated_at)
-  VALUES ('analysis.scheduled',
-        'Scheduled Analysis',
-        'This configuration parameter allows setting a schedule for analysis. When enabled, analysis will only run when the scheduled time arrives',
-        '{"enabled": false, "rrule": ""}',
-        current_timestamp,current_timestamp) ON CONFLICT DO NOTHING;
+-- Backfill our sso_providers table with the existing data from saml_providers
+-- The hardcoded type is determined by the AuthProvider
+-- See:
+-- https://github.com/SpecterOps/BloodHound/blob/main/cmd/api/src/model/auth.go#L565
+INSERT INTO sso_providers(name, slug, type) (SELECT name, lower(replace(name, ' ', '-')), 1
+                                             FROM saml_providers
+                                             WHERE sso_provider_id IS NULL)
+ON CONFLICT DO NOTHING;
 
--- Add last analysis time to datapipe status so we can track scheduled analysis time properly
-ALTER TABLE datapipe_status
-ADD COLUMN IF NOT EXISTS "last_analysis_run_at" TIMESTAMP with time zone;
+-- Backfill the references from the newly created sso_provider entries
+UPDATE saml_providers
+SET sso_provider_id = (SELECT id FROM sso_providers WHERE name = saml_providers.name)
+WHERE saml_providers.sso_provider_id IS NULL;
+
+-- Add the sso_provider to the users table
+ALTER TABLE ONLY users
+    ADD COLUMN IF NOT EXISTS sso_provider_id INTEGER NULL;
+ALTER TABLE ONLY users
+    DROP CONSTRAINT IF EXISTS fk_users_sso_provider;
+ALTER TABLE ONLY users
+    ADD CONSTRAINT fk_users_sso_provider FOREIGN KEY (sso_provider_id) REFERENCES sso_providers (id) ON DELETE CASCADE;
+
+-- Backfill users with their proper sso_provider when they have a saml_provider_id
+UPDATE users u
+SET sso_provider_id = (SELECT sso.id
+                       FROM saml_providers saml
+                                JOIN sso_providers sso ON sso.id = saml.sso_provider_id
+                       WHERE u.saml_provider_id = saml.id)
+WHERE sso_provider_id IS NULL
+  AND saml_provider_id IS NOT NULL;
```

Comment:
While it's likely blows up scope, we might want to put deprecated on the struct for the saml_provider_id. Idk how much extra work it would be to swap over to that in the logic or not but something to audit before merging this. If it's minimal, we might want to consider it as well as dropping the column from this table.... 🤷

Reply:
There are already a ton of moving parts. I vote deal with it in a separate ticket.